aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/SmokeTestsDecorator.kt

```diff
@@ -0,0 +1,172 @@
+    override val name: String = "SmokeTests"
+    override val order: Byte = 0
+
+    override fun operationCustomizations(
+        codegenContext: ClientCodegenContext,
+        operation: OperationShape,
+        baseCustomizations: List<OperationCustomization>,
+    ): List<OperationCustomization> =
+        baseCustomizations.letIf(operation.hasTrait(SmokeTestsTrait.ID)) {
+            it + SmokeTestsOperationCustomization(codegenContext, operation)
+        }
+}
+
+class SmokeTestsOperationCustomization(
+    private val codegenContext: ClientCodegenContext,
+    private val operation: OperationShape,
+) : OperationCustomization() {
+    private val smokeTestsTrait = operation.expectTrait<SmokeTestsTrait>()
+    private val testCases: List<SmokeTestCase> = smokeTestsTrait.testCases
+    private val operationInput = operation.inputShape(codegenContext.model)
+
+    override fun section(section: OperationSection): Writable {
+        if (testCases.isEmpty()) return emptySection
+
+        return when (section) {
+            is OperationSection.UnitTests ->
+                writable {
+                    testCases.forEach { testCase ->
+                        Attribute(cfg("v2SmokeTests")).render(this)
+                        Attribute.TokioTest.render(this)
+                        this.rustBlock("async fn test_${testCase.id.toSnakeCase()}()") {
+                            val instantiator = SmokeTestsInstantiator(codegenContext)
+                            instantiator.renderConf(this, testCase.vendorParams)
+                            rust("let client = crate::Client::from_conf(conf);")
+                            instantiator.renderInput(this, operation, operationInput, testCase.params)
+                            instantiator.renderExpectation(this, testCase.expectation)
+                        }
+                    }
+                }
+
+            else -> emptySection
+        }
+    }
+}
+
+class SmokeTestsBuilderKindBehavior(val codegenContext: CodegenContext) : Instantiator.BuilderKindBehavior {
+    override fun hasFallibleBuilder(shape: StructureShape): Boolean =
+        BuilderGenerator.hasFallibleBuilder(shape, codegenContext.symbolProvider)
+
+    override fun setterName(memberShape: MemberShape): String = memberShape.setterName()
+
+    override fun doesSetterTakeInOption(memberShape: MemberShape): Boolean = true
+}
+
+class SmokeTestsInstantiator(private val codegenContext: ClientCodegenContext) : Instantiator(
+    codegenContext.symbolProvider,
+    codegenContext.model,
+    codegenContext.runtimeConfig,
+    SmokeTestsBuilderKindBehavior(codegenContext),
+) {
+    fun renderConf(
+        writer: RustWriter,
+        data: Optional<ObjectNode>,
+    ) {
+        writer.rust("let conf = crate::config::Builder::new()")
+        writer.indent()
+        writer.rust(".behavior_version(crate::config::BehaviorVersion::latest())")
+        data.orNull()?.let { node ->
```

Comment:
I don't think you need to parse all of these yourself, you should be able to use [AwsVendorParams](https://github.com/smithy-lang/smithy/blob/6466fe77c65b8a17b219f0b0a60c767915205f95/smithy-aws-smoke-test-model/src/main/java/software/amazon/smithy/aws/smoketests/model/AwsVendorParams.java#L14) (or the S3VendorParams) that comes from `smithy-aws-smoke-test-model` package.

```kt
// made up API for Optional/ObjectNode but you get the idea
val params = AwsVendorParams(data.getOr(ObjectNode::empty())

writer.rust(".use_dual_stack(${params.useDualStack()}")
...
```

Reply:
`AwsVendorParams` can't be constructed afaict. `init` is private.

Reply:
You're right, init is private, looks like you can use [this](https://github.com/smithy-lang/smithy/blob/6466fe77c65b8a17b219f0b0a60c767915205f95/smithy-aws-smoke-test-model/src/main/java/software/amazon/smithy/aws/smoketests/model/AwsSmokeTestModel.java) to create an instance though 🤷‍♂️